File: android/app/src/main/kotlin/com/example/ai_chatbot/OverlayMemory.kt
package com.example.ai_chatbot

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * Isolated overlay conversation memory — never mixed with main app chat.
 */
object OverlayMemory {
    private const val PREFS = "aura_overlay_memory"
    private const val KEY_HISTORY = "history_json"
    private const val MAX_TURNS = 8

    fun load(context: Context): JSONArray {
        val raw = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getString(KEY_HISTORY, "[]") ?: "[]"
        return try {
            JSONArray(raw)
        } catch (_: Exception) {
            JSONArray()
        }
    }

    fun save(context: Context, history: JSONArray) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putString(KEY_HISTORY, history.toString())
            .apply()
    }

    fun appendTurn(context: Context, user: String, assistant: String) {
        val history = load(context)
        history.put(JSONObject().apply {
            put("role", "user")
            put("content", user)
        })
        history.put(JSONObject().apply {
            put("role", "assistant")
            put("content", assistant)
        })
        while (history.length() > MAX_TURNS * 2) {
            history.remove(0)
        }
        save(context, history)
    }

    fun clear(context: Context) {
        save(context, JSONArray())
    }
}

File: lib/overlay_runtime/overlay_memory.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

class OverlayMemory {
  static const String _keyHistory = 'aura_overlay_history_dart';
  static const int maxTurns = 10;

  /// Loads history turns as a List of Maps
  static Future<List<Map<String, String>>> load() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final String? raw = prefs.getString(_keyHistory);
      if (raw == null || raw.isEmpty) return [];
      
      final List<dynamic> decoded = jsonDecode(raw);
      return decoded.map((item) {
        return {
          'role': item['role']?.toString() ?? '',
          'content': item['content']?.toString() ?? '',
        };
      }).toList();
    } catch (e) {
      print("OverlayMemory failed to load: $e");
      return [];
    }
  }

  /// Saves the complete list of turns
  static Future<void> save(List<Map<String, String>> history) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(_keyHistory, jsonEncode(history));
    } catch (e) {
      print("OverlayMemory failed to save: $e");
    }
  }

  /// Appends a new user & assistant message turn and keeps it within the limit
  static Future<void> appendTurn(String userText, String assistantText) async {
    final List<Map<String, String>> history = await load();
    history.add({'role': 'user', 'content': userText});
    history.add({'role': 'assistant', 'content': assistantText});
    
    // limit history size
    while (history.length > maxTurns * 2) {
      history.removeAt(0);
    }
    await save(history);
  }

  /// Wipes history
  static Future<void> clear() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.remove(_keyHistory);
    } catch (e) {
      print("OverlayMemory failed to clear: $e");
    }
  }
}
